contracts/src/token/erc1155/extensions/uri_storage.rs

```diff
@@ -0,0 +1,222 @@
+//! ERC-1155 token with storage based token URI management.
+//!
+//! Inspired by the [`crate::token::erc721::extensions::Erc721UriStorage`]
+use alloc::string::String;
+
+use alloy_primitives::U256;
+use stylus_sdk::{evm, stylus_proc::sol_storage};
+
+use super::metadata_uri::{IErc1155MetadataUri, URI};
+
+sol_storage! {
+    /// Uri Storage.
+    pub struct Erc1155UriStorage {
+        /// Optional base URI
+        string _base_uri;
+        /// Optional mapping for token URIs.
+        mapping(uint256 => string) _token_uris;
+    }
+}
+
+impl Erc1155UriStorage {
+    /// Returns the Uniform Resource Identifier (URI) for `token_id` token.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    /// * `token_id` - Id of a token.
+    /// * `metadata_uri` - Read access to a contract providing
+    ///   [`IErc1155MetadataUri`] interface.
+    ///
+    /// NOTE: In order to have [`Erc1155UriStorage::uri`] exposed in ABI,
+    /// you need to do this manually.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    ///     pub fn uri(&self, token_id: U256) -> String {
+    ///         self.uri_storage.uri(token_id, &self.metadata_uri)
+    ///     }
+    pub fn uri(
+        &self,
+        token_id: U256,
+        metadata_uri: &impl IErc1155MetadataUri,
+    ) -> String {
+        let token_uri = self._token_uris.get(token_id).get_string();
+
+        let uri = if token_uri.is_empty() {
+            metadata_uri.uri(token_id)
+        } else {
+            self._base_uri.get_string() + &token_uri
+        };
+
+        uri
+    }
+
+    /// Sets `token_uri` as the tokenURI of `token_id`.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `token_id` - Id of a token.
+    /// * `token_uri` - URI for the token.
+    /// * `metadata_uri` - Read access to a contract providing
+    ///   [`IErc1155MetadataUri`] interface.
+    ///
+    /// # Events
+    ///
+    /// Emits a [`URI`] event.
+    pub fn set_token_uri(
+        &mut self,
+        token_id: U256,
+        token_uri: String,
+        metadata_uri: &impl IErc1155MetadataUri,
+    ) {
+        self._token_uris.setter(token_id).set_str(token_uri);
+        evm::log(URI { value: self.uri(token_id, metadata_uri), id: token_id });
+    }
+
+    /// Sets `base_uri` as the `_base_uri` for all tokens.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `base_uri` - New base URI.
+    pub fn set_base_uri(&mut self, base_uri: String) {
+        self._base_uri.set_str(base_uri);
+    }
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
+    use alloy_primitives::U256;
+    use stylus_sdk::stylus_proc::sol_storage;
+
+    use super::Erc1155UriStorage;
+    use crate::token::erc1155::{extensions::Erc1155MetadataUri, Erc1155};
+
+    fn random_token_id() -> U256 {
```

Comment:
I copied this helper fn from other unit tests, but I think it's unnecessary in all of them, as the tests are simple enough. A predetermined token ID would allow us to remove [`rand` as a dependency in _/contracts/_](https://github.com/OpenZeppelin/rust-contracts-stylus/blob/main/contracts/Cargo.toml#L24).

Reply:
Tracking issue https://github.com/OpenZeppelin/rust-contracts-stylus/issues/435

Reply:
Good idea!